File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long timeTaken(vector<int>& diffs , vector<int>& times , int level){
    int size = diffs.size();
    long long taken = 0;

    for(int i = 0; i < size; i++) {
        if(diffs[i] <= level) {
            taken += times[i]; 
        } else {
            int cnt = diffs[i] - level;  // 틀린 횟수
            int prevTime = (i == 0) ? 0 : times[i - 1];
            long long failCost = (long long)(times[i] + prevTime) * cnt;
            taken += failCost + times[i];

            // 제한 시간을 초과하면 더 볼 필요 없음
            if(taken > 1e15) return taken;
        }
    }

    return taken;
}

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 0;
    int left = 1; 
    int right = *max_element(diffs.begin(), diffs.end());

    while(left <= right) {
        int mid = (left + right) / 2;
        long long taken = timeTaken(diffs, times, mid);

        if(taken <= limit) {
            answer = mid;    
            right = mid - 1;
        } else {
            left = mid + 1;    
        }
    }

    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;

class Solution {
    public long solution(int[] diffs, int[] times, long limit) {
        long left = 1;
        long right = Arrays.stream(diffs).max().getAsInt();
        long answer = right + 1;

        while (left <= right) {
            long mid = (left + right) / 2;

            if (isFinished(mid, diffs, times, limit)) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return answer;
    }

    private boolean isFinished(long level, int[] diffs, int[] times, long limit) {
        int n = diffs.length;
        long totalTime = 0;

        for (int i = 0; i < n; ++i) {
            if (diffs[i] <= level) {
                totalTime += times[i];
            } else {
                long mistake = diffs[i] - level;
                long penalty = (i == 0 ? 0 : times[i - 1]);
                long timeCur = times[i];
                totalTime += mistake * (timeCur + penalty) + timeCur;
            }

            if (totalTime > limit) return false; 
        }

        return true;
    }
}